pkg/assembler/backends/testing/backend.go

```diff
@@ -16,12 +16,43 @@
 package testing
 
 import (
+	"sync"
+
 	"github.com/guacsec/guac/pkg/assembler/backends"
 	"github.com/guacsec/guac/pkg/assembler/graphql/model"
 )
 
+var (
+	index      = indexType{}
+	packages   = pkgTypeMap{}
+	sources    = srcTypeMap{}
+	sourceMaps = srcMaps{}
+)
+
 type DemoCredentials struct{}
 
+// IDs: We have a global ID for all nodes that have references to/from.
+// Since we always ingest data and never remove, we can keep this global and
+// increment it as needed.
+// For fast retrieval, we also keep a map from ID from nodes that have it.
+// IDs are stored as string in graphql even though we ask for integers
+// See https://github.com/99designs/gqlgen/issues/2561
+type nodeID int
+
+type hasID interface {
+	getID() nodeID
+}
+
+type indexType map[nodeID]hasID
+
+// locking to ensure ID is not duplicated
+func (c *demoClient) getNextID() nodeID {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.id = c.id + 1
```

Comment:
For this, probably can use atomic add instead of a lock. https://pkg.go.dev/sync/atomic#Int32.Add